File: scala/src/main/scala/com/spoonacular/IngredientsApi.scala
package com.spoonacular

import java.io._
import spoonacular._
import com.spoonacular.client.model._
import java.math.BigDecimal
import com.spoonacular.client.model.InlineObject2
import com.spoonacular.client.model.InlineResponse20022
import com.spoonacular.client.model.InlineResponse20024
import com.spoonacular.client.model.InlineResponse20025
import com.spoonacular.client.model.InlineResponse20026
import com.spoonacular.client.model.InlineResponse20034
import com.spoonacular.client.model.RecipesParseIngredientsNutritionWeightPerServing
import io.finch.circe._
import io.circe.generic.semiauto._
import com.twitter.concurrent.AsyncStream
import com.twitter.finagle.Service
import com.twitter.finagle.Http
import com.twitter.finagle.http.{Request, Response}
import com.twitter.finagle.http.exp.Multipart.{FileUpload, InMemoryFileUpload, OnDiskFileUpload}
import com.twitter.util.Future
import com.twitter.io.Buf
import io.finch._, items._
import java.io.File
import java.time._

object IngredientsApi {
    /**
    * Compiles all service endpoints.
    * @return Bundled compilation of all service endpoints.
    */
    def endpoints(da: DataAccessor) =
        autocompleteIngredientSearch(da) :+:
        computeIngredientAmount(da) :+:
        getIngredientInformation(da) :+:
        getIngredientSubstitutes(da) :+:
        getIngredientSubstitutesByID(da) :+:
        ingredientSearch(da) :+:
        ingredientsByIDImage(da) :+:
        mapIngredientsToGroceryProducts(da) :+:
        visualizeIngredients(da)


    private def checkError(e: CommonError) = e match {
      case InvalidInput(_) => BadRequest(e)
      case MissingIdentifier(_) => BadRequest(e)
      case RecordNotFound(_) => NotFound(e)
      case _ => InternalServerError(e)
    }

    implicit class StringOps(s: String) {

      import java.time.format.DateTimeFormatter

      lazy val localformatter: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")
      lazy val datetimeformatter: DateTimeFormatter =
      DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSSZ")

      def toLocalDateTime: LocalDateTime = LocalDateTime.parse(s,localformatter)
      def toZonedDateTime: ZonedDateTime = ZonedDateTime.parse(s, datetimeformatter)

    }

        /**
        * 
        * @return An endpoint representing a Seq[InlineResponse20024]
        */
        private def autocompleteIngredientSearch(da: DataAccessor): Endpoint[Seq[InlineResponse20024]] =
        get("food" :: "ingredients" :: "autocomplete" :: paramOption("query") :: paramOption("number").map(_.map(_.toInt)) :: paramOption("metaInformation").map(_.map(_.toBoolean)) :: paramOption("intolerances") :: param("apiKey")) { (query: Option[String], number: Option[Int], metaInformation: Option[Boolean], intolerances: Option[String], authParamapiKeyScheme: String) =>
          da.Ingredients_autocompleteIngredientSearch(query, number, metaInformation, intolerances, authParamapiKeyScheme) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }

        /**
        * 
        * @return An endpoint representing a RecipesParseIngredientsNutritionWeightPerServing
        */
        private def computeIngredientAmount(da: DataAccessor): Endpoint[RecipesParseIngredientsNutritionWeightPerServing] =
        get("food" :: "ingredients" :: bigdecimal :: "amount" :: param("nutrient") :: param("target").map(_.toBigDecimal) :: paramOption("unit") :: param("apiKey")) { (id: BigDecimal, nutrient: String, target: BigDecimal, unit: Option[String], authParamapiKeyScheme: String) =>
          da.Ingredients_computeIngredientAmount(id, nutrient, target, unit, authParamapiKeyScheme) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }

        /**
        * 
        * @return An endpoint representing a InlineResponse20022
        */
        private def getIngredientInformation(da: DataAccessor): Endpoint[InlineResponse20022] =
        get("food" :: "ingredients" :: int :: "information" :: paramOption("amount").map(_.map(_.toBigDecimal)) :: paramOption("unit") :: param("apiKey")) { (id: Int, amount: Option[BigDecimal], unit: Option[String], authParamapiKeyScheme: String) =>
          da.Ingredients_getIngredientInformation(id, amount, unit, authParamapiKeyScheme) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }

        /**
        * 
        * @return An endpoint representing a InlineResponse20026
        */
        private def getIngredientSubstitutes(da: DataAccessor): Endpoint[InlineResponse20026] =
        get("food" :: "ingredients" :: "substitutes" :: param("ingredientName") :: param("apiKey")) { (ingredientName: String, authParamapiKeyScheme: String) =>
          da.Ingredients_getIngredientSubstitutes(ingredientName, authParamapiKeyScheme) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }

        /**
        * 
        * @return An endpoint representing a InlineResponse20026
        */
        private def getIngredientSubstitutesByID(da: DataAccessor): Endpoint[InlineResponse20026] =
        get("food" :: "ingredients" :: int :: "substitutes" :: param("apiKey")) { (id: Int, authParamapiKeyScheme: String) =>
          da.Ingredients_getIngredientSubstitutesByID(id, authParamapiKeyScheme) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }

        /**
        * 
        * @return An endpoint representing a InlineResponse20025
        */
        private def ingredientSearch(da: DataAccessor): Endpoint[InlineResponse20025] =
        get("food" :: "ingredients" :: "search" :: paramOption("query") :: paramOption("addChildren").map(_.map(_.toBoolean)) :: paramOption("minProteinPercent").map(_.map(_.toBigDecimal)) :: paramOption("maxProteinPercent").map(_.map(_.toBigDecimal)) :: paramOption("minFatPercent").map(_.map(_.toBigDecimal)) :: paramOption("maxFatPercent").map(_.map(_.toBigDecimal)) :: paramOption("minCarbsPercent").map(_.map(_.toBigDecimal)) :: paramOption("maxCarbsPercent").map(_.map(_.toBigDecimal)) :: paramOption("metaInformation").map(_.map(_.toBoolean)) :: paramOption("intolerances") :: paramOption("sort") :: paramOption("sortDirection") :: paramOption("offset").map(_.map(_.toInt)) :: paramOption("number").map(_.map(_.toInt)) :: param("apiKey")) { (query: Option[String], addChildren: Option[Boolean], minProteinPercent: Option[BigDecimal], maxProteinPercent: Option[BigDecimal], minFatPercent: Option[BigDecimal], maxFatPercent: Option[BigDecimal], minCarbsPercent: Option[BigDecimal], maxCarbsPercent: Option[BigDecimal], metaInformation: Option[Boolean], intolerances: Option[String], sort: Option[String], sortDirection: Option[String], offset: Option[Int], number: Option[Int], authParamapiKeyScheme: String) =>
          da.Ingredients_ingredientSearch(query, addChildren, minProteinPercent, maxProteinPercent, minFatPercent, maxFatPercent, minCarbsPercent, maxCarbsPercent, metaInformation, intolerances, sort, sortDirection, offset, number, authParamapiKeyScheme) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }

        /**
        * 
        * @return An endpoint representing a Object
        */
        private def ingredientsByIDImage(da: DataAccessor): Endpoint[Object] =
        get("recipes" :: bigdecimal :: "ingredientWidget.png" :: paramOption("measure") :: param("apiKey")) { (id: BigDecimal, measure: Option[String], authParamapiKeyScheme: String) =>
          da.Ingredients_ingredientsByIDImage(id, measure, authParamapiKeyScheme) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }

        /**
        * 
        * @return An endpoint representing a Seq[InlineResponse20034]
        */
        private def mapIngredientsToGroceryProducts(da: DataAccessor): Endpoint[Seq[InlineResponse20034]] =
        post("food" :: "ingredients" :: "map" :: jsonBody[InlineObject2] :: param("apiKey")) { (inlineObject2: InlineObject2, authParamapiKeyScheme: String) =>
          da.Ingredients_mapIngredientsToGroceryProducts(inlineObject2, authParamapiKeyScheme) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }

        /**
        * 
        * @return An endpoint representing a String
        */
        private def visualizeIngredients(da: DataAccessor): Endpoint[String] =
        post("recipes" :: "visualizeIngredients" :: headerOption("Content-Type") :: paramOption("language") :: headerOption("Accept") :: param("apiKey")) { (contentType: Option[String], language: Option[String], accept: Option[String], authParamapiKeyScheme: String) =>
          da.Ingredients_visualizeIngredients(contentType, language, accept, authParamapiKeyScheme) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }


    implicit private def fileUploadToFile(fileUpload: FileUpload) : File = {
      fileUpload match {
        case upload: InMemoryFileUpload =>
          bytesToFile(Buf.ByteArray.Owned.extract(upload.content))
        case upload: OnDiskFileUpload =>
          upload.content
        case _ => null
      }
    }

    private def bytesToFile(input: Array[Byte]): java.io.File = {
      val file = File.createTempFile("tmpIngredientsApi", null)
      val output = new FileOutputStream(file)
      output.write(input)
      file
    }

    // This assists in params(string) application (which must be Seq[A] in parameter list) when the param is used as a List[A] elsewhere.
    implicit def seqList[A](input: Seq[A]): List[A] = input.toList
}

File: dart/com/spoonacular/ingredients_api.dart
part of openapi.api;



class IngredientsApi {
  final ApiClient apiClient;

  IngredientsApi([ApiClient apiClient]) : apiClient = apiClient ?? defaultApiClient;

  /// Autocomplete Ingredient Search
  ///
  /// Autocomplete the entry of an ingredient.
  Future<List<InlineResponse20024>> autocompleteIngredientSearch({ String query, int number, bool metaInformation, String intolerances }) async {
    Object postBody;

    // verify required params are set

    // create path and map variables
    String path = "/food/ingredients/autocomplete".replaceAll("{format}","json");

    // query params
    List<QueryParam> queryParams = [];
    Map<String, String> headerParams = {};
    Map<String, String> formParams = {};
    if(query != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "query", query));
    }
    if(number != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "number", number));
    }
    if(metaInformation != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "metaInformation", metaInformation));
    }
    if(intolerances != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "intolerances", intolerances));
    }

    List<String> contentTypes = [];

    String contentType = contentTypes.isNotEmpty ? contentTypes[0] : "application/json";
    List<String> authNames = ["apiKeyScheme"];

    if(contentType.startsWith("multipart/form-data")) {
      bool hasFields = false;
      MultipartRequest mp = new MultipartRequest(null, null);
      if(hasFields)
        postBody = mp;
    }
    else {
    }

    var response = await apiClient.invokeAPI(path,
                                             'GET',
                                             queryParams,
                                             postBody,
                                             headerParams,
                                             formParams,
                                             contentType,
                                             authNames);

    if(response.statusCode >= 400) {
      throw new ApiException(response.statusCode, _decodeBodyBytes(response));
    } else if(response.body != null) {
      return (apiClient.deserialize(_decodeBodyBytes(response), 'List<InlineResponse20024>') as List).map((item) => item as InlineResponse20024).toList();
    } else {
      return null;
    }
  }
  /// Compute Ingredient Amount
  ///
  /// Compute the amount you need of a certain ingredient for a certain nutritional goal. For example, how much pineapple do you have to eat to get 10 grams of protein?
  Future<RecipesParseIngredientsNutritionWeightPerServing> computeIngredientAmount(num id, String nutrient, num target, { String unit }) async {
    Object postBody;

    // verify required params are set
    if(id == null) {
     throw new ApiException(400, "Missing required param: id");
    }
    if(nutrient == null) {
     throw new ApiException(400, "Missing required param: nutrient");
    }
    if(target == null) {
     throw new ApiException(400, "Missing required param: target");
    }

    // create path and map variables
    String path = "/food/ingredients/{id}/amount".replaceAll("{format}","json").replaceAll("{" + "id" + "}", id.toString());

    // query params
    List<QueryParam> queryParams = [];
    Map<String, String> headerParams = {};
    Map<String, String> formParams = {};
      queryParams.addAll(_convertParametersForCollectionFormat("", "nutrient", nutrient));
      queryParams.addAll(_convertParametersForCollectionFormat("", "target", target));
    if(unit != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "unit", unit));
    }

    List<String> contentTypes = [];

    String contentType = contentTypes.isNotEmpty ? contentTypes[0] : "application/json";
    List<String> authNames = ["apiKeyScheme"];

    if(contentType.startsWith("multipart/form-data")) {
      bool hasFields = false;
      MultipartRequest mp = new MultipartRequest(null, null);
      if(hasFields)
        postBody = mp;
    }
    else {
    }

    var response = await apiClient.invokeAPI(path,
                                             'GET',
                                             queryParams,
                                             postBody,
                                             headerParams,
                                             formParams,
                                             contentType,
                                             authNames);

    if(response.statusCode >= 400) {
      throw new ApiException(response.statusCode, _decodeBodyBytes(response));
    } else if(response.body != null) {
      return apiClient.deserialize(_decodeBodyBytes(response), 'RecipesParseIngredientsNutritionWeightPerServing') as RecipesParseIngredientsNutritionWeightPerServing;
    } else {
      return null;
    }
  }
  /// Get Ingredient Information
  ///
  /// Use an ingredient id to get all available information about an ingredient, such as its image and supermarket aisle.
  Future<InlineResponse20022> getIngredientInformation(int id, { num amount, String unit }) async {
    Object postBody;

    // verify required params are set
    if(id == null) {
     throw new ApiException(400, "Missing required param: id");
    }

    // create path and map variables
    String path = "/food/ingredients/{id}/information".replaceAll("{format}","json").replaceAll("{" + "id" + "}", id.toString());

    // query params
    List<QueryParam> queryParams = [];
    Map<String, String> headerParams = {};
    Map<String, String> formParams = {};
    if(amount != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "amount", amount));
    }
    if(unit != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "unit", unit));
    }

    List<String> contentTypes = [];

    String contentType = contentTypes.isNotEmpty ? contentTypes[0] : "application/json";
    List<String> authNames = ["apiKeyScheme"];

    if(contentType.startsWith("multipart/form-data")) {
      bool hasFields = false;
      MultipartRequest mp = new MultipartRequest(null, null);
      if(hasFields)
        postBody = mp;
    }
    else {
    }

    var response = await apiClient.invokeAPI(path,
                                             'GET',
                                             queryParams,
                                             postBody,
                                             headerParams,
                                             formParams,
                                             contentType,
                                             authNames);

    if(response.statusCode >= 400) {
      throw new ApiException(response.statusCode, _decodeBodyBytes(response));
    } else if(response.body != null) {
      return apiClient.deserialize(_decodeBodyBytes(response), 'InlineResponse20022') as InlineResponse20022;
    } else {
      return null;
    }
  }
  /// Get Ingredient Substitutes
  ///
  /// Search for substitutes for a given ingredient.
  Future<InlineResponse20026> getIngredientSubstitutes(String ingredientName) async {
    Object postBody;

    // verify required params are set
    if(ingredientName == null) {
     throw new ApiException(400, "Missing required param: ingredientName");
    }

    // create path and map variables
    String path = "/food/ingredients/substitutes".replaceAll("{format}","json");

    // query params
    List<QueryParam> queryParams = [];
    Map<String, String> headerParams = {};
    Map<String, String> formParams = {};
      queryParams.addAll(_convertParametersForCollectionFormat("", "ingredientName", ingredientName));

    List<String> contentTypes = [];

    String contentType = contentTypes.isNotEmpty ? contentTypes[0] : "application/json";
    List<String> authNames = ["apiKeyScheme"];

    if(contentType.startsWith("multipart/form-data")) {
      bool hasFields = false;
      MultipartRequest mp = new MultipartRequest(null, null);
      if(hasFields)
        postBody = mp;
    }
    else {
    }

    var response = await apiClient.invokeAPI(path,
                                             'GET',
                                             queryParams,
                                             postBody,
                                             headerParams,
                                             formParams,
                                             contentType,
                                             authNames);

    if(response.statusCode >= 400) {
      throw new ApiException(response.statusCode, _decodeBodyBytes(response));
    } else if(response.body != null) {
      return apiClient.deserialize(_decodeBodyBytes(response), 'InlineResponse20026') as InlineResponse20026;
    } else {
      return null;
    }
  }
  /// Get Ingredient Substitutes by ID
  ///
  /// Search for substitutes for a given ingredient.
  Future<InlineResponse20026> getIngredientSubstitutesByID(int id) async {
    Object postBody;

    // verify required params are set
    if(id == null) {
     throw new ApiException(400, "Missing required param: id");
    }

    // create path and map variables
    String path = "/food/ingredients/{id}/substitutes".replaceAll("{format}","json").replaceAll("{" + "id" + "}", id.toString());

    // query params
    List<QueryParam> queryParams = [];
    Map<String, String> headerParams = {};
    Map<String, String> formParams = {};

    List<String> contentTypes = [];

    String contentType = contentTypes.isNotEmpty ? contentTypes[0] : "application/json";
    List<String> authNames = ["apiKeyScheme"];

    if(contentType.startsWith("multipart/form-data")) {
      bool hasFields = false;
      MultipartRequest mp = new MultipartRequest(null, null);
      if(hasFields)
        postBody = mp;
    }
    else {
    }

    var response = await apiClient.invokeAPI(path,
                                             'GET',
                                             queryParams,
                                             postBody,
                                             headerParams,
                                             formParams,
                                             contentType,
                                             authNames);

    if(response.statusCode >= 400) {
      throw new ApiException(response.statusCode, _decodeBodyBytes(response));
    } else if(response.body != null) {
      return apiClient.deserialize(_decodeBodyBytes(response), 'InlineResponse20026') as InlineResponse20026;
    } else {
      return null;
    }
  }
  /// Ingredient Search
  ///
  /// Search for simple whole foods (e.g. fruits, vegetables, nuts, grains, meat, fish, dairy etc.).
  Future<InlineResponse20025> ingredientSearch({ String query, bool addChildren, num minProteinPercent, num maxProteinPercent, num minFatPercent, num maxFatPercent, num minCarbsPercent, num maxCarbsPercent, bool metaInformation, String intolerances, String sort, String sortDirection, int offset, int number }) async {
    Object postBody;

    // verify required params are set

    // create path and map variables
    String path = "/food/ingredients/search".replaceAll("{format}","json");

    // query params
    List<QueryParam> queryParams = [];
    Map<String, String> headerParams = {};
    Map<String, String> formParams = {};
    if(query != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "query", query));
    }
    if(addChildren != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "addChildren", addChildren));
    }
    if(minProteinPercent != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "minProteinPercent", minProteinPercent));
    }
    if(maxProteinPercent != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "maxProteinPercent", maxProteinPercent));
    }
    if(minFatPercent != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "minFatPercent", minFatPercent));
    }
    if(maxFatPercent != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "maxFatPercent", maxFatPercent));
    }
    if(minCarbsPercent != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "minCarbsPercent", minCarbsPercent));
    }
    if(maxCarbsPercent != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "maxCarbsPercent", maxCarbsPercent));
    }
    if(metaInformation != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "metaInformation", metaInformation));
    }
    if(intolerances != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "intolerances", intolerances));
    }
    if(sort != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "sort", sort));
    }
    if(sortDirection != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "sortDirection", sortDirection));
    }
    if(offset != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "offset", offset));
    }
    if(number != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "number", number));
    }

    List<String> contentTypes = [];

    String contentType = contentTypes.isNotEmpty ? contentTypes[0] : "application/json";
    List<String> authNames = ["apiKeyScheme"];

    if(contentType.startsWith("multipart/form-data")) {
      bool hasFields = false;
      MultipartRequest mp = new MultipartRequest(null, null);
      if(hasFields)
        postBody = mp;
    }
    else {
    }

    var response = await apiClient.invokeAPI(path,
                                             'GET',
                                             queryParams,
                                             postBody,
                                             headerParams,
                                             formParams,
                                             contentType,
                                             authNames);

    if(response.statusCode >= 400) {
      throw new ApiException(response.statusCode, _decodeBodyBytes(response));
    } else if(response.body != null) {
      return apiClient.deserialize(_decodeBodyBytes(response), 'InlineResponse20025') as InlineResponse20025;
    } else {
      return null;
    }
  }
  /// Ingredients by ID Image
  ///
  /// Visualize a recipe&#39;s ingredient list.
  Future<Object> ingredientsByIDImage(num id, { String measure }) async {
    Object postBody;

    // verify required params are set
    if(id == null) {
     throw new ApiException(400, "Missing required param: id");
    }

    // create path and map variables
    String path = "/recipes/{id}/ingredientWidget.png".replaceAll("{format}","json").replaceAll("{" + "id" + "}", id.toString());

    // query params
    List<QueryParam> queryParams = [];
    Map<String, String> headerParams = {};
    Map<String, String> formParams = {};
    if(measure != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "measure", measure));
    }

    List<String> contentTypes = [];

    String contentType = contentTypes.isNotEmpty ? contentTypes[0] : "application/json";
    List<String> authNames = ["apiKeyScheme"];

    if(contentType.startsWith("multipart/form-data")) {
      bool hasFields = false;
      MultipartRequest mp = new MultipartRequest(null, null);
      if(hasFields)
        postBody = mp;
    }
    else {
    }

    var response = await apiClient.invokeAPI(path,
                                             'GET',
                                             queryParams,
                                             postBody,
                                             headerParams,
                                             formParams,
                                             contentType,
                                             authNames);

    if(response.statusCode >= 400) {
      throw new ApiException(response.statusCode, _decodeBodyBytes(response));
    } else if(response.body != null) {
      return apiClient.deserialize(_decodeBodyBytes(response), 'Object') as Object;
    } else {
      return null;
    }
  }
  /// Map Ingredients to Grocery Products
  ///
  /// Map a set of ingredients to products you can buy in the grocery store.
  Future<List<InlineResponse20034>> mapIngredientsToGroceryProducts(InlineObject2 inlineObject2) async {
    Object postBody = inlineObject2;

    // verify required params are set
    if(inlineObject2 == null) {
     throw new ApiException(400, "Missing required param: inlineObject2");
    }

    // create path and map variables
    String path = "/food/ingredients/map".replaceAll("{format}","json");

    // query params
    List<QueryParam> queryParams = [];
    Map<String, String> headerParams = {};
    Map<String, String> formParams = {};

    List<String> contentTypes = ["application/json"];

    String contentType = contentTypes.isNotEmpty ? contentTypes[0] : "application/json";
    List<String> authNames = ["apiKeyScheme"];

    if(contentType.startsWith("multipart/form-data")) {
      bool hasFields = false;
      MultipartRequest mp = new MultipartRequest(null, null);
      if(hasFields)
        postBody = mp;
    }
    else {
    }

    var response = await apiClient.invokeAPI(path,
                                             'POST',
                                             queryParams,
                                             postBody,
                                             headerParams,
                                             formParams,
                                             contentType,
                                             authNames);

    if(response.statusCode >= 400) {
      throw new ApiException(response.statusCode, _decodeBodyBytes(response));
    } else if(response.body != null) {
      return (apiClient.deserialize(_decodeBodyBytes(response), 'List<InlineResponse20034>') as List).map((item) => item as InlineResponse20034).toList();
    } else {
      return null;
    }
  }
  /// Ingredients Widget
  ///
  /// Visualize ingredients of a recipe. You can play around with that endpoint!
  Future<String> visualizeIngredients({ String contentType, String language, String accept }) async {
    Object postBody;

    // verify required params are set

    // create path and map variables
    String path = "/recipes/visualizeIngredients".replaceAll("{format}","json");

    // query params
    List<QueryParam> queryParams = [];
    Map<String, String> headerParams = {};
    Map<String, String> formParams = {};
    if(language != null) {
      queryParams.addAll(_convertParametersForCollectionFormat("", "language", language));
    }
    headerParams["Content-Type"] = contentType;
    headerParams["Accept"] = accept;

    List<String> contentTypes = ["application/x-www-form-urlencoded"];

    String contentType = contentTypes.isNotEmpty ? contentTypes[0] : "application/json";
    List<String> authNames = ["apiKeyScheme"];

    if(contentType.startsWith("multipart/form-data")) {
      bool hasFields = false;
      MultipartRequest mp = new MultipartRequest(null, null);
      if(hasFields)
        postBody = mp;
    }
    else {
    }

    var response = await apiClient.invokeAPI(path,
                                             'POST',
                                             queryParams,
                                             postBody,
                                             headerParams,
                                             formParams,
                                             contentType,
                                             authNames);

    if(response.statusCode >= 400) {
      throw new ApiException(response.statusCode, _decodeBodyBytes(response));
    } else if(response.body != null) {
      return apiClient.deserialize(_decodeBodyBytes(response), 'String') as String;
    } else {
      return null;
    }
  }
}
